File: backend/notification-service/src/main/java/com/employee/notificationservice/service/NotificationService.java
package com.employee.notificationservice.service;

import com.employee.notificationservice.entity.Notification;
import com.employee.notificationservice.repository.NotificationRepository;
import jakarta.annotation.PostConstruct;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Service
@RequiredArgsConstructor
@Slf4j
public class NotificationService {

    private final NotificationRepository notificationRepository;

    @PostConstruct
    public void seedInitialNotifications() {
        if (notificationRepository.count() == 0) {
            log.info("Seeding initial enterprise notifications into notification_db...");

            notificationRepository.save(Notification.builder()
                    .recipientId(1L)
                    .recipientEmail("admin@company.com")
                    .title("Task Assigned: Configure Spring Cloud Gateway Ingress")
                    .message("You have been assigned to Jira Issue #1: Configure Spring Cloud Gateway Ingress & Microservices Routing. High Priority.")
                    .category("TASK")
                    .priority("HIGH")
                    .status("SENT")
                    .readStatus(false)
                    .createdAt(LocalDateTime.now().minusHours(2))
                    .sentAt(LocalDateTime.now().minusHours(2))
                    .build());

            notificationRepository.save(Notification.builder()
                    .recipientId(1L)
                    .recipientEmail("admin@company.com")
                    .title("Annual Performance Evaluation Published")
                    .message("Your Q3 AI Performance Appraisal review has been generated and published by HR leadership.")
                    .category("HR")
                    .priority("MEDIUM")
                    .status("SENT")
                    .readStatus(false)
                    .createdAt(LocalDateTime.now().minusHours(5))
                    .sentAt(LocalDateTime.now().minusHours(5))
                    .build());

            notificationRepository.save(Notification.builder()
                    .recipientId(1L)
                    .recipientEmail("admin@company.com")
                    .title("Automated Task Duration Log Completed")
                    .message("Task #4 (Fix CORS Preflight Headers) moved to DONE. 3.5 elapsed work hours automatically synchronized into ledger.")
                    .category("ALERT")
                    .priority("MEDIUM")
                    .status("SENT")
                    .readStatus(false)
                    .createdAt(LocalDateTime.now().minusHours(12))
                    .sentAt(LocalDateTime.now().minusHours(12))
                    .build());

            notificationRepository.save(Notification.builder()
                    .recipientId(1L)
                    .recipientEmail("admin@company.com")
                    .title("Eureka Microservice Health Check Passed")
                    .message("All 8 Spring Cloud microservices registered cleanly with Eureka Service Registry.")
                    .category("SYSTEM")
                    .priority("LOW")
                    .status("SENT")
                    .readStatus(true)
                    .createdAt(LocalDateTime.now().minusDays(1))
                    .sentAt(LocalDateTime.now().minusDays(1))
                    .build());

            notificationRepository.save(Notification.builder()
                    .recipientId(1L)
                    .recipientEmail("admin@company.com")
                    .title("Failed Event Webhook Dispatch Alert")
                    .message("Simulated delivery attempt to external audit webhook failed. Automatic retry queued.")
                    .category("ALERT")
                    .priority("HIGH")
                    .status("FAILED")
                    .errorMessage("HTTP 503 Service Unavailable - Endpoint Timeout (3000ms)")
                    .readStatus(false)
                    .createdAt(LocalDateTime.now().minusDays(2))
                    .sentAt(LocalDateTime.now().minusDays(2))
                    .build());
        }
    }

    public List<Notification> getUserNotifications(Long userId, String category, Boolean unreadOnly) {
        String cat = (category != null && !category.trim().isEmpty() && !"ALL".equalsIgnoreCase(category)) ? category.trim().toUpperCase() : null;
        return notificationRepository.searchUserNotifications(userId, cat, unreadOnly);
    }

    public List<Notification> getRecentActivityStream() {
        return notificationRepository.findTop10ByOrderByCreatedAtDesc();
    }


    public Long getUnreadCount(Long userId) {
        return notificationRepository.countByRecipientIdAndReadStatusFalse(userId);
    }

    @Transactional
    public Notification markAsRead(Long id) {
        Notification notification = notificationRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Notification not found with id: " + id));
        notification.setReadStatus(true);
        return notificationRepository.save(notification);
    }

    @Transactional
    public void markAllAsRead(Long userId) {
        notificationRepository.markAllAsReadForUser(userId);
    }

    @Transactional
    public void deleteNotification(Long id) {
        notificationRepository.deleteById(id);
    }

    @Transactional
    public Notification dispatchNotification(Notification notification) {
        if (notification.getCreatedAt() == null) {
            notification.setCreatedAt(LocalDateTime.now());
        }
        if (notification.getSentAt() == null) {
            notification.setSentAt(LocalDateTime.now());
        }
        if (notification.getStatus() == null) {
            notification.setStatus("SENT");
        }
        return notificationRepository.save(notification);
    }

    public Map<String, Object> getDeliveryMetrics() {
        Map<String, Object> metrics = new HashMap<>();
        long sent = notificationRepository.countByStatus("SENT");
        long pending = notificationRepository.countByStatus("PENDING");
        long failed = notificationRepository.countByStatus("FAILED");
        long total = notificationRepository.count();

        metrics.put("sentCount", sent);
        metrics.put("pendingCount", pending);
        metrics.put("failedCount", failed);
        metrics.put("totalCount", total);
        metrics.put("deliverySuccessRate", total > 0 ? Math.round(((double) sent / total) * 100.0) : 100);
        return metrics;
    }
}

File: frontend/src/app/services/notification.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { BehaviorSubject, Observable, interval, of } from 'rxjs';
import { catchError, map, switchMap } from 'rxjs/operators';
import { environment } from '../../environments/environment';

export interface NotificationItem {
  id?: number;
  recipientId: number;
  recipientEmail?: string;
  title: string;
  message: string;
  category: 'TASK' | 'SYSTEM' | 'HR' | 'ALERT';
  priority: 'LOW' | 'MEDIUM' | 'HIGH';
  status?: 'PENDING' | 'SENT' | 'FAILED';
  errorMessage?: string;
  readStatus?: boolean;
  createdAt?: string;
  sentAt?: string;
}

export interface NotificationMetrics {
  sentCount: number;
  pendingCount: number;
  failedCount: number;
  totalCount: number;
  deliverySuccessRate: number;
}

interface ApiResponse<T> {
  message: string;
  data: T;
  timeStamp: string;
}

@Injectable({
  providedIn: 'root',
})
export class NotificationService {
  private baseUrl = `${environment.apiUrl}/api/notifications`;

  private unreadCountSubject = new BehaviorSubject<number>(3);
  public unreadCount$ = this.unreadCountSubject.asObservable();

  constructor(private http: HttpClient) {
    // Auto-polling every 10 seconds to update unread badge counter
    interval(10000)
      .pipe(
        switchMap(() => this.getUnreadCount(1).pipe(catchError(() => of({ data: 3 } as any))))
      )
      .subscribe((res) => {
        if (res && res.data !== undefined) {
          this.unreadCountSubject.next(res.data);
        }
      });
  }

  getUserNotifications(userId: number, category?: string, unreadOnly?: boolean): Observable<ApiResponse<NotificationItem[]>> {
    let params = new HttpParams();
    if (category && category !== 'ALL') params = params.set('category', category);
    if (unreadOnly) params = params.set('unreadOnly', 'true');

    return this.http.get<ApiResponse<NotificationItem[]>>(`${this.baseUrl}/user/${userId}`, { params });
  }

  getUnreadCount(userId: number): Observable<ApiResponse<number>> {
    return this.http.get<ApiResponse<number>>(`${this.baseUrl}/user/${userId}/unread-count`).pipe(
      map((res) => {
        if (res && res.data !== undefined) {
          this.unreadCountSubject.next(res.data);
        }
        return res;
      })
    );
  }

  markAsRead(id: number): Observable<ApiResponse<NotificationItem>> {
    return this.http.patch<ApiResponse<NotificationItem>>(`${this.baseUrl}/${id}/read`, {}).pipe(
      map((res) => {
        const current = this.unreadCountSubject.value;
        if (current > 0) this.unreadCountSubject.next(current - 1);
        return res;
      })
    );
  }

  markAllAsRead(userId: number): Observable<ApiResponse<void>> {
    return this.http.patch<ApiResponse<void>>(`${this.baseUrl}/user/${userId}/read-all`, {}).pipe(
      map((res) => {
        this.unreadCountSubject.next(0);
        return res;
      })
    );
  }

  deleteNotification(id: number): Observable<ApiResponse<void>> {
    return this.http.delete<ApiResponse<void>>(`${this.baseUrl}/${id}`);
  }

  dispatchNotification(notification: NotificationItem): Observable<ApiResponse<NotificationItem>> {
    return this.http.post<ApiResponse<NotificationItem>>(`${this.baseUrl}/dispatch`, notification);
  }

  getMetrics(): Observable<ApiResponse<NotificationMetrics>> {
    return this.http.get<ApiResponse<NotificationMetrics>>(`${this.baseUrl}/metrics`);
  }
}
